src/functions/fast-login.ts

```diff
@@ -0,0 +1,233 @@
+ */
+export const getAssertionUserIdVsCfVerifier = (
+  fiscalCodeFromHeader: FiscalCode
+): Verifier => (assertionDoc): ReturnType<Verifier> =>
+  pipe(
+    assertionDoc,
+    getFiscalNumberFromSamlResponse,
+    TE.fromOption(
+      () =>
+        new H.HttpError(
+          "Missing or invalid Fiscal Code in the retrieved saml assertion."
+        )
+    ),
+    TE.filterOrElse(
+      fiscalCodeFromAssertion =>
+        fiscalCodeFromAssertion === fiscalCodeFromHeader,
+      fiscalCodeFromAssertion =>
+        new H.HttpError(
+          `The provided user id do not match the fiscalNumber in the assertion: fromSaml=${fiscalCodeFromAssertion},fromHeader=${fiscalCodeFromHeader}`
+        )
+    ),
+    TE.map(() => true as const)
+  );
+
+export const makeFastLoginHandler: H.Handler<
+  H.HttpRequest,
+  H.HttpResponse<FastLoginSAML, 200>,
+  FnLollipopClientDependency
+> = H.of((req: H.HttpRequest) =>
+  pipe(
+    req,
+    sequenceS(RTE.ApplyPar)({
+      lollipopHeaders: RequiredHeadersMiddleware(LollipopHeaders),
+      publicKey: RequiredHeaderMiddleware(
+        JwkPublicKeyFromToken,
+        PUBLIC_KEY_HEADER_NAME
+      )
+    }),
+    TE.chainFirstW(verifiedHeaders =>
+      pipe(
+        [
+          req,
+          verifiedHeaders.lollipopHeaders[ASSERTION_REF_HEADER_NAME],
+          verifiedHeaders.publicKey
+        ] as const,
+        params =>
+          pipe(
+            validateHttpSignatureWithEconding("der")(...params),
+            TE.orElse(() =>
+              validateHttpSignatureWithEconding("ieee-p1363")(...params)
+            )
+          ),
+        TE.mapLeft(
+          () => new H.HttpUnauthorizedError("Invalid Lollipop Signature")
+        )
+      )
+    ),
+    RTE.fromTaskEither,
+    rte =>
+      pipe(
+        rte,
+        RTE.chain(_ => RetrieveSAMLResponse(_.lollipopHeaders)),
+        RTE.chainFirstW(({ saml_response }) =>
+          pipe(
+            TE.tryCatch(
+              async () =>
+                new DOMParser().parseFromString(saml_response, "text/xml"),
+              () => new H.HttpError("Error parsing the SAMLResponse")
+            ),
+            RTE.fromTaskEither,
+            RTE.bindTo("samlResponseDoc"),
+            RTE.bindW("verifiedHeaders", () => rte),
+            RTE.chain(_ =>
+              pipe(
+                [
+                  getAssertionRefVsInRensponseToVerifier(
+                    _.verifiedHeaders.publicKey,
+                    _.verifiedHeaders.lollipopHeaders[ASSERTION_REF_HEADER_NAME]
+                  ),
+                  getAssertionUserIdVsCfVerifier(
+                    _.verifiedHeaders.lollipopHeaders[
+                      "x-pagopa-lollipop-user-id"
+                    ]
+                  )
+                ],
+                RA.map(verifier => verifier(_.samlResponseDoc)),
+                TE.sequenceArray,
+                RTE.fromTaskEither
+              )
+            )
+          )
+        )
+      ),
```

Comment:
Can we make it a little more linear?

```suggestion
    RTE.bindTo("verifiedHeaders"),
    RTE.bindW("samlResponse", ({ verifiedHeaders }) =>
      RetrieveSAMLResponse(verifiedHeaders.lollipopHeaders)
    ),
    RTE.bindW("samlResponseDoc", ({ samlResponse }) =>
      pipe(
        TE.tryCatch(
          async () =>
            new DOMParser().parseFromString(
              samlResponse.saml_response,
              "text/xml"
            ),
          () => new H.HttpError("Error parsing the SAMLResponse")
        ),
        RTE.fromTaskEither
      )
    ),
    RTE.chainFirstW(({ verifiedHeaders, samlResponseDoc }) =>
      pipe(
        [
          getAssertionRefVsInRensponseToVerifier(
            verifiedHeaders.publicKey,
            verifiedHeaders.lollipopHeaders[ASSERTION_REF_HEADER_NAME]
          ),
          getAssertionUserIdVsCfVerifier(
            verifiedHeaders.lollipopHeaders["x-pagopa-lollipop-user-id"]
          )
        ],
        RA.map(verifier => verifier(samlResponseDoc)),
        TE.sequenceArray,
        RTE.fromTaskEither
      )
    ),
    RTE.map(({ samlResponse }) => samlResponse),
```

Reply:
Fixed in f37570c